\documentclass[11pt]{article}

\usepackage{fullpage}
\usepackage[round,semicolon]{natbib}
\usepackage{amssymb}
\usepackage{amsmath}
\usepackage{bbm}
\usepackage{graphicx}
\usepackage{hyperref}

\DeclareMathOperator{\erf}{erf}
\DeclareMathOperator{\lngamma}{\ln\!\Gamma}

% NOTE
% one sentence per line for nice git diffs
% WD: I'm a fan of initialed source comments like this

\title{Coalescent inference of mutation spectrum histories from sample frequency spectra}

\author{
W. DeWitt$^{1}$, K. Harris$^{2}$, and K. Harris$^{1}$\\
\small{Departments of $^1$Genome Sciences and $^2$Biology, University of Washington, Seattle, USA}
% \small{$^\ast$ Equal contribution}
}

\begin{document}

\maketitle

\begin{abstract}

Single nucleotide variant (SNV) frequencies partitioned according to triplet nucleotide context vary among human ancestry groups and among great ape species, indicating variation and divergence of the mutation process.
The sample frequency spectrum (SFS)---the distribution of derived allele frequencies among sampled haplotypes---is a well-studied population genetic summary statistic that is sensitive to demographic history.
We extend a coalescent framework for demographic inference from the SFS to accommodate inference of mutation spectrum histories from the triplet-resolved sample frequency spectrum (3-SFS).
We formulate inference of mutation spectrum histories from the 3-SFS as a linear inverse problem.

\end{abstract}


\section*{Introduction}\label{sec:intro}

\cite{Harris2017-fw} showed triplet SNV spectrum differences between human groups and between great apes, and used a coalescent simulation approach to fit the timing of a pulse of \texttt{T\underline{C}C>T\underline{T}C} mutations in Europeans.
We develop coalescent theory for the dependence of the expected 3-SFS on both demographic history (effective population size as a function of time into the past) \emph{and} triplet mutation spectrum history (triplet-specific mutation intensities as a function of time into the past).
This is similar in spirit to Kelley's simulation-based approach (which assumed the European demography of \cite{Tennessen2012-dq}), but is much faster, more flexibly parameterized.

Some references for nonidentifiability of demographic history from the SFS: \cite{Baharian2018-np, Bhaskar2014-fu, Terhorst2015-xt, Myers2008-jp}.
It might be good to consider identifiability of mutation intensity history from the SFS conditioned on demographic history.

%WD todo: inverse problem intro, forward Vs inverse problems, and ill-posedness. Point out \eta is nonlinear, but \mu is linear

%WD todo: the scope of this paper is \mu inference conditioned on \eta. There are plenty of fancy methods one can use to infer \eta

\section*{Model}\label{sec:model}

The setting for our modeling is Kingman's coalescent \citep{Kingman1982-ge, Kingman1982-tf, Kingman1982-ys, Kingman2000-jr}, with all the usual niceties: neutrality, infinite sites, linkage equilibrium, and panmixia.
In the Methods we detail our retracing of the derivation by \cite{Griffiths1998-qf} of the frequency distribution of a derived allele conditioned on the demographic history, while generalizing to a time inhomogeneous mutation process.
We make use of the results of \cite{Polanski2003-kg} and \cite{Polanski2003-ll} to facilitate computation, and build on the notation of \cite{Rosen2018-bb} for a finite-dimensional approximation of our demographic and mutation intensity histories.
Complete derivation of formulae are found in the Appendix.

Given $n$ sampled haplotypes and a demographic history $\eta(t)$ (with time measured retrospectively from the present), we show in Appendix \ref{sec:appendix:xi} that the expected SFS $\boldsymbol \xi = (\xi_1, \xi_2,\dots, \xi_{n-1})$ is a linear transform of the mutation intensity history $\mu(t)$:
\begin{equation}
  \label{eqn:transform}
\boldsymbol \xi = \mathcal{L}_{n,\eta}\mu,
\end{equation}
where $\mathcal{L}_{n,\eta}$ is a finite-rank linear operator that maps infinite-dimensional mutation intensity histories to $(n-1)$-dimensional SFS vectors.
It is dependent on $n$ and $\eta$, which we take to be fixed.

% Need to talk about this fig \ref{fig:model}.

\begin{figure}
\centering
\includegraphics[width=\textwidth]{figures/model}
\caption{Schematic of the marked Poisson process model with $n=4$.
We condition on coalescent times $T_4=t_4,T_3=t_3,T_2=t_2$ and consider mutation intensity function $\mu(t)$.
Red dots indicate mutation events placed by time $t$, genomic position $x$, and coalescent line (which are depicted as extruded in the genomic coordinate axis, grey sheets).
The probability that a differential element $dxdt$ on a given sheet contains a mutation is proportional to the instantaneous mutation intensity $\mu(t)$.
}
\label{fig:model}
\end{figure}

For numerical implementation we consider finite-dimensional approximations of $\eta(t)$ and $\mu(t)$ as piecewise constant on $m$ common epochs $[t_0, t_1), [t_1, t_2),\dots, [t_{m-1}, t_m)$ where $0=t_0 < t_1 < \dots < t_{m-1} < t_m=\infty$.
We take the epoch boundaries as fixed parameters, and in practice make them dense so as to approximate infinite-dimensional histories.
Let $\boldsymbol y = (y_1,\dots,y_m)$ denote the constant population size $\eta(t)$ during each epoch, and let $\boldsymbol z = (z_1,\dots,z_m)$ denote the constant mutation intensity $\mu(t)$ during each epoch.
In Appendix \ref{sec:appendix:pcsws} we show the linear transform \eqref{eqn:transform} reduces to the matrix equation
\begin{equation}
\label{eqn:transform_discrete}
\boldsymbol \xi = L_{n, \boldsymbol y} \boldsymbol z,
\end{equation}
where the $(n-1)\times m$ matrix $L_{n, \boldsymbol y}$ is fixed given a fixed demographic history $\boldsymbol y$.

In linkage equilibrium the log likelihood function of the mutation intensity history $\boldsymbol z$ given an observed SFS vector $\boldsymbol x = (x_1,\dots,x_{n-1})$ and demographic history $\boldsymbol y$ is given by the Poisson random field approximation \citep{}
\[
\ell(\boldsymbol z; \boldsymbol x, \boldsymbol y) = \boldsymbol x^\intercal\log(L_{n, \boldsymbol y} \boldsymbol z) - \boldsymbol 1^\intercal L_{n, \boldsymbol y} \boldsymbol z,
\]
where $\log(\cdot)$ operates elementwise, and we've dropped a constant term wrt $\boldsymbol z$.

%WD more inverse problem intro
The inverse problem \eqref{eqn:transform} is ill-posed in general, so many very different histories can be equally consistent with the data \citep{oscillation paper? Yun's other papers?}.
To penalize oscillatory histories we regularize on the derivative of $\mu$.
We define $L_p^p$ derivative regularizers for history $\mu(t)$:
\[
\int_0^\infty\left|\frac{d\mu(t)}{dt}\right|^p dt.
\]
In the finite-dimensional approximation \eqref{eqn:transform_discrete} the derivatives become Dirac delta functions at the change points, and this become
\[
\left|\Delta\boldsymbol z\right|_p^p.
\]
where $\Delta$ denotes the first difference operator.
Our penalized log-likelihood is then
\begin{equation}
\label{eqn:penalized}
\tilde\ell(\boldsymbol z) = \ell(\boldsymbol z) - \lambda \left(\alpha\left|\Delta\boldsymbol z\right|_1 + \frac{(1-\alpha)}{2}\left|\Delta\boldsymbol z\right|_2^2\right),
\end{equation}
where $\lambda$ tunes the regularization strength, and $\alpha$ tunes the relative $p=1$ vs $p=2$ regularization.


\section*{Results}\label{sec:results}

% Todo:
% \begin{itemize}
% \item repeat 1KG analysis like in \cite{Harris2017-fw}, see if we recapitulate Kelley's simulation-based pulse results
% \item cluster triplet time series to see if we pull in minor components.
% \end{itemize}

\begin{figure}
  \centering
  \includegraphics[width=.7\textwidth]{figures/fit_teaser}
  \caption{Simulating and inverting a pulse.}
  \label{}
\end{figure}

\subsection*{Tempora incognita: observability toward the coalescent horizon}\label{sec:model:loss}

Todo: SVD on $L_{n, \boldsymbol y}$

Intuitively, we know the SFS can't contain any information about the history beyond the TMRCA $T_2$, since mutations that occurred before then will not be segregating in the sample.
Thus, we will find it useful to penalize complexity in the history more heavily at times that are more probably ancestral to the TMRCA.

From \eqref{eqn:pi} and \eqref{eqn:r} in the Methods, the CDF of $T_2$ is
\begin{align}
F_2(t) &= 1 - \sum_{j=2}^n A_{2,j}r_2(t)\\
&= 1 - \sum_{j=2}^n A_{2,j}r_2(t)
\end{align}


\section*{Discussion}\label{sec:discussion}

Joint inference of $\eta$ and $\mu$.


\section*{Methods}\label{sec:methods}

\subsection*{Implementation and pipeline}\label{sec:methods:tool}

The \texttt{mushi} software package is available at \url{https://github.com/harrispopgen/mushi}.
%WD cite whatever packages we use, e.g. proxtv

\section*{Acknowledgements}\label{sec:ack}

\bibliographystyle{plainnat}
\bibliography{refs}


\appendix
\input{appendix}


\end{document}
